libbeat/publisher/queue/memqueue/produce.go

```diff
@@ -42,66 +41,83 @@ type openState struct {
 	events chan pushRequest
 }
 
+// producerID stores the order of events within a single producer, so multiple
+// event acknowledgement callbacks can be coalesced into a single call.
+// It is defined as an explicit type to reduce cross-confusion with the id
+// of an event in the queue itself, which is a queue.EntryID.
+// with the
+type producerID uint64
+
 type produceState struct {
 	cb        ackHandler
-	dropCB    func(beat.Event)
+	dropCB    func(interface{})
 	cancelled bool
-	lastACK   uint32
+	lastACK   producerID
 }
 
 type ackHandler func(count int)
 
-func newProducer(b *broker, cb ackHandler, dropCB func(beat.Event), dropOnCancel bool) queue.Producer {
+func newProducer(b *broker, cb ackHandler, dropCB func(interface{}), dropOnCancel bool) queue.Producer {
 	openState := openState{
 		log:    b.logger,
 		done:   make(chan struct{}),
 		events: b.pushChan,
 	}
 
 	if cb != nil {
-		p := &ackProducer{broker: b, seq: 1, dropOnCancel: dropOnCancel, openState: openState}
+		p := &ackProducer{broker: b, dropOnCancel: dropOnCancel, openState: openState}
 		p.state.cb = cb
 		p.state.dropCB = dropCB
 		return p
 	}
 	return &forgetfulProducer{broker: b, openState: openState}
 }
 
-func (p *forgetfulProducer) Publish(event interface{}) bool {
-	return p.openState.publish(pushRequest{event: event})
+func (p *forgetfulProducer) makePushRequest(event interface{}) pushRequest {
+	resp := make(chan queue.EntryID, 1)
```

Comment:
We make a channel for each event? Is it worth making one for each batch? Could we make a version of `Publish` that accepts an entire event batch instead of an individual event?

Reply:
It's not impossible but it would significantly complicate things: right now the memory queue accepts publish requests iff it has at least one free event slot, because then it can definitely handle the request without blocking. Accepting multiple events per push request would mean that's no longer true, so we'd need to assemble the response over time, and both event loops would need more complicated state handling to spread the response across multiple iterations.

However, ephemeral channels are quite cheap and this pattern is already used heavily in both queues. (We could, though, probably cache a single response channel for each producer, since the producers themselves are only supposed to be used from a single goroutine?)

Reply:
Thanks, the pattern from the shipper side is to try to publish entire batches instead of individual events. I was more interested in seeing if we could improve the API for that use case, but it sounds like it isn't worth it unless there is some evidence of a performance issue.

Reply:
Ah, in that case yes I'd say it's a possible future optimization but probably not worth the complexity until we see where the real bottlenecks are. (My current belief is that the channeled API calls are negligible compared to the actual queueing process, so our current approach of calling single-event publish in a for loop doesn't lose much. It also gives us free semi-fair balancing between different inputs -- in the simplest implementation, keeping published batches together for the queue call would block all other inputs while a large publish request was being processed.)